### DISCUSSION TITLE
Multi level hierarchy with Serialization/Deserialization and configuration value

### DISCUSSION THREAD

**Discussion by micsport13** [FasterXML/jackson-databind#4465] [Q&A]
I'm creating a fake data generator project and as a result, I have created this existing interface:
```
@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, property = "type")
@JsonSubTypes({
  @JsonSubTypes.Type(value = BigIntDataType.class, name = "bigint"),
  @JsonSubTypes.Type(value = BinaryDataType.class, name = "binary"),
  @JsonSubTypes.Type(
      value = BitDataType.class,
      names = {"bit", "boolean"}),
  @JsonSubTypes.Type(value = CharDataType.class, name = "char"),
  @JsonSubTypes.Type(value = DateDataType.class, name = "date"),
  @JsonSubTypes.Type(value = DateTime2DataType.class, name = "datetime2"),
  @JsonSubTypes.Type(value = DateTimeOffsetDataType.class, name = "datetimeoffset"),
  @JsonSubTypes.Type(value = DecimalDataType.class, name = "decimal"),
  @JsonSubTypes.Type(value = FloatDataType.class, name = "float"),
  @JsonSubTypes.Type(
      value = IntegerDataType.class,
      names = {"integer", "int"}),
  @JsonSubTypes.Type(value = MoneyDataType.class, name = "money"),
  @JsonSubTypes.Type(value = NVarcharDataType.class, name = "nvarchar"),
  @JsonSubTypes.Type(value = RealDataType.class, name = "real"),
  @JsonSubTypes.Type(value = SmallIntDataType.class, name = "smallint"),
  @JsonSubTypes.Type(value = SmallMoneyDataType.class, name = "smallmoney"),
  @JsonSubTypes.Type(value = TextDataType.class, name = "text"),
  @JsonSubTypes.Type(value = TimeDataType.class, name = "time"),
  @JsonSubTypes.Type(value = TinyIntDataType.class, name = "tinyint"),
  @JsonSubTypes.Type(value = UniqueIdentifierDataType.class, name = "uniqueidentifier"),
  @JsonSubTypes.Type(value = VarcharDataType.class, name = "varchar"),
})
public interface DataType<T> {

  T cast(Object value) throws MismatchedDataTypeException;
}
```
Which works well for me, however, I'm looking to make my program a little more modular in terms of converting all of these subtypes into abstract classes and then creating implementations based off of the abstract classes that are currently normal classes (like DecimalDataType).  One example would be DecimalDataType is subclassed by MySQLDecimalDataType, PostgresDecimalDataType, SQLServerDataType, etc.  My question is, is there a way that I can pass an environment variable to know that a DecimalDataType should get deserialized into a PostgresDecimalDataType if the variable DB_ENGINE=Postgres for example, similar to an abstract factory pattern?  I would like to keep them serialized the same but get deserialized into a concrete class.  If so, how would one do this?  If not, do I have to create a custom deserializer for each one of these classes and just switch based on the deserialization context?  Any ideas would be greatly appreciated.

**Comment by yawkat** [comment]
you can register subtypes manually on ObjectMapper

**Comment by micsport13** [comment]
If I want them all to get serialized the same way but then get deserialized into the concrete class, does that change anything?  There shouldn't be much difference between the concrete classes and the abstract base classes.
